File: fichier_matlab/circ_data.m
function [R1, R1_b, theta,V,v,delta,s,k]= circ_data(vec)
%% Prend vec un vecteur d'angles en degrés et renvoie :
% R1 : resultant length (longueur résultate)
% R1_b : mean resultant length (la moyenne de la longueur résultante)
% theta : mean direction (la direction moyenne) (estimation)
% V : circular variance (variance circulaire)
% v : circular standard deviantion (variance circulaire standard)
% delta : circular dispertion (dispertion circulaire, pour les calculs
% d'intervalles de confiance)
% s : skewness
% k : kurtosis
N=length(vec);

%% Appels aux fonctions de création pour Cp, Cp_barre, etc.
C1=f_C(vec,N,1);
C2=f_C(vec,N,2);
C1_b=C1/N;
C2_b=C2/N;

S1=f_S(vec,N,1);
S2=f_S(vec,N,2);
S1_b=S1/N;
S2_b=S2/N;

T1=f_T(C1_b,S1_b);
T2=f_T(C2_b,S2_b);

%% Calculs des retours
%R :
R1=sqrt(C1^2+S1^2);
R2=sqrt(C2^2+S2^2);

R1_b=R1/N;
R2_b=R2/N;

%theta :
theta=T1;

%V et v :
V=1-R1_b;
v=sqrt(-2*log(R1_b));

%delta :
delta=(1-T2)/(2*R1_b^2);

%s :
s=(R2_b*sin(T2-2*T1))/(1-R1_b)^(3/2);

%k :
k=(R2_b*cos(T2-2*T1)-R1_b^4)/(1-R1_b)^2;
end

File: fichier_C/parametre/circ_data.c
#include "circ_data.h"

float circ_data_Sin(float* vec,int taille,int p)
{
	float Sp=0;
	for (int i = 0; i < taille; ++i)
	{
		float temp=p*vec[i];
		Sp=Sp+sin(temp);
	}
	return Sp;
}

float circ_data_Cos(float* vec,int taille,int p)
{
	float Cp=0;
	for (int i = 0; i < taille; ++i)
	{
		Cp=Cp+cos(p*vec[i]);
	}
	return Cp;
}

float circ_data_ATan(float Cp_b, float Sp_b)
{
	float Tp=atan(Sp_b/Cp_b);

	if (Cp_b<0)
	{
		Tp=Tp+3.14;
	}
	else if(Cp_b>0 && Sp_b<0)
	{
		Tp=Tp+2*3.14;
	}

	return Tp;
}

float circ_data_R(float C, float S)
{
	return sqrt(pow(C,2)+pow(S,2));
};

float circ_data_Var(float R)
{
	return 1-R;
}

float circ_data_sigma(float R)
{
	return sqrt(-2*logf(R));
}

float circ_data_delta(float T, float R)
{
	return (1-T)/(2*R*R);
}

float circ_data_skewnessis(float R1, float R2, float T1, float T2)
{
	return (R2*sin(T2-2*T1))/pow((1-R1),3/2);
}

float circ_data_kurtosis(float R1, float R2, float T1, float T2)
{
	return (R2*cos(T2-2*T1)-pow(R1,4))/pow((1-R1),3/2);
}

void circ_data_calcul(float* vec, int taille, float* R1, float* R1_b, float* theta, float* Var, float* sigma, float* delta, float* skewnessis, float* kurtosis)
{

	//Creation Cp, Cp_b, Sp, Sp_b 
	float C1=circ_data_Cos(vec,taille,1);
	float C2=circ_data_Cos(vec,taille,2);
	float C1_b=C1/(float)taille;
	float C2_b=C2/(float)taille;

	float S1=circ_data_Sin(vec,taille,1);
	float S2=circ_data_Sin(vec,taille,2);
	float S1_b=S1/(float)taille;
	float S2_b=S2/(float)taille;

	float T1=circ_data_ATan(C1_b,S1_b);
	float T2=circ_data_ATan(C2_b,S2_b);

	//Calculs des retours
	*(R1)=circ_data_R(C1,S1);
	float R2=circ_data_R(C2,S2);

	*(R1_b)=*(R1)/(float)taille;
	float R2_b=R2/(float)taille;	

	*(theta)=T1;

	*(Var)=circ_data_Var(*(R1_b));

	*(sigma)=circ_data_sigma(*(R1_b));

	*(delta)=circ_data_delta(T2,*(R1_b));

	*(skewnessis)=circ_data_skewnessis(*(R1_b),R2_b,T1,T2);

	*(kurtosis)=circ_data_kurtosis(*(R1_b),R2_b,T1,T2);
};